src/tests/system/tests/test_ipa.py

```diff
@@ -640,3 +640,393 @@ def test_ipa__switch_user_with_smartcard_authentication(client: Client, ipa: IPA
     assert (
         "ipacertuser1" in result.stdout
     ), f"'ipacertuser1' not found in 'whoami' output! Stdout content: {result.stdout}"
+
+
+@pytest.mark.importance("medium")
+@pytest.mark.topology(KnownTopology.IPA)
+def test_ipa__hbac_permitted_users_can_login(client: Client, ipa: IPA):
+    """
+    :title: Validate HBAC rule-based SSH access control using the `sshd` service
+    :setup:
+      1. Add users
+      2. Disable any existing HBAC rules
+      3. Create an HBAC rule granting a user access
+    :steps:
+      1. Restart SSSD and attempt SSH login as users
+      2. Remove the HBAC rule, restart SSSD, and attempt SSH login again
+    :expectedresults:
+      1. Permitted user gains SSH access
+      2. All users are denied SSH access after rule removal
+    :customerscenario: False
+    """
+    # Expected failed logins during testing might trigger the per-source penalty protection inside of sshd
+    # so we exempt local SSH connections (::1, 127.0.0.1) from per-source connection penalties
+    client.sshd.config_set(
+        [
+            {
+                "PerSourcePenaltyExemptList": "::1,127.0.0.1",
+            }
+        ]
+    )
+    client.sshd.reload()
+
+    users = ["user1", "user2", "user3"]
+    for user in users:
+        ipa.user(user).add()
+    ipa.hbac("allow_all").disable()
+    ssh_access_rule = ipa.hbac("ssh_access_user1").create(
+        description="SSH access rule for user1", users=["user1"], hosts=["client.test"], services=["sshd"]
+    )
+
+    client.sssd.restart()
+    assert client.auth.ssh.password("user1", "Secret123"), "user1 SSH should succeed!"
+    assert not client.auth.ssh.password("user2", "Secret123"), "user2 SSH should be denied!"
+    assert not client.auth.ssh.password("user3", "Secret123"), "user3 SSH should be denied!"
+
+    ssh_access_rule.delete()
+    client.sssd.restart()
+    for user in users:
+        assert not client.auth.ssh.password(user, "Secret123"), f"{user} SSH should not succeed!"
+
+
+@pytest.mark.importance("medium")
+@pytest.mark.topology(KnownTopology.IPA)
+def test_ipa__hbac_permitted_group_users_can_login(client: Client, ipa: IPA):
+    """
+    :title: Validate HBAC rule-based access control using user groups
+    :setup:
+      1. Add users and create user groups
+      2. Assign users to groups
+      3. Disable default HBAC rule
+      4. Create HBAC rule for a group SSH access.
+    :steps:
+      1. Remove a user from the group, restart SSSD
+      2. Attempt to login again
+    :expectedresults:
+      1. User is removed and SSSD has restarted
+      2. The remaining group member is permitted to login
+    :customerscenario: False
+    """
+    u1 = ipa.user("user1").add()
+    u2 = ipa.user("user2").add()
+    u3 = ipa.user("user3").add()
```

Comment:
Add a line break for each step. 

Add the users code
<empty line>
Assign users to groups code
<empty line>

etc